R/get_stan_data.R

```diff
@@ -42,68 +42,230 @@ get_input_ww_data_for_stan <- function(preprocessed_ww_data,
+#' Get mapping from lab-site to site
+#'
+#' @param input_ww_data a dataframe of the wastewater data to be passed
+#' directly to stan, must have the following columns: date, site, lab,
+#' genome_copies_per_ml, site_pop, below_lod, and exclude
+#'
+#' @return a dataframe mapping the unique combinations of sites and labs
+#' to their indices in the model and the population of the site in that
+#' observation unit (lab_site)
+#' @export
+#'
+get_lab_site_site_spine <- function(input_ww_data) {
+  ww_data_present <- !is.null(input_ww_data)
+
+  if (ww_data_present) {
+    lab_site_site_spine <-
+      input_ww_data |>
+      dplyr::select(
+        "lab_site_index", "site_index",
+        "site", "lab", "site_pop"
+      ) |>
+      dplyr::arrange(.data$lab_site_index) |>
+      dplyr::distinct() |>
+      dplyr::mutate(
+        "lab_site_name" = glue::glue(
+          "Site: {site}, Lab: {lab}"
+        )
+      )
+  } else {
+    lab_site_site_spine <- tibble::tibble()
+  }
 
-  ww_data_sizes <- get_ww_data_sizes(
-    ww_data,
-    lod_col_name = "below_lod"
-  )
 
-  ww_indices <- get_ww_data_indices(
-    ww_data,
-    first_count_data_date,
-    owt = ww_data_sizes$owt,
-    lod_col_name = "below_lod"
-  )
+  return(lab_site_site_spine)
+}
 
-  ww_data <- ww_data |>
-    dplyr::mutate(
-      t = ww_indices$ww_sampled_times
+#' Get site to subpopulation map
+#'
+#' @param input_ww_data a dataframe of the wastewater data to be passed
+#' directly to stan, must have the following columns: date, site, lab,
+#' genome_copies_per_ml, site_pop, below_lod, and exclude
+#' @param input_count_data a dataframe of the count data to be passed
+#' directly to stan, , must have the following columns: date, count, total_pop
+#'
+#' @return a dataframe mapping the sites to the corresponding subpopulation and
+#' subpopulation index, plus the population in each subpopulation. Imposes
+#' the logic to add a subpopulation if the total population is greater than
+#' the sum of the site populations in the input wastewater data
+#' @export
+#'
+get_site_subpop_spine <- function(input_ww_data,
+                                  input_count_data) {
+  ww_data_present <- !is.null(input_ww_data)
+
+  total_pop <- input_count_data |>
+    dplyr::distinct(.data$total_pop) |>
+    dplyr::pull()
+
+  if (ww_data_present) {
+    add_auxiliary_subpop <- ifelse(
+      total_pop > sum(unique(input_ww_data$site_pop)),
+      TRUE,
+      FALSE
     )
+    site_indices <- input_ww_data |>
+      dplyr::select("site_index", "site", "site_pop") |>
+      dplyr::distinct() |>
+      dplyr::arrange(.data$site_index)
 
-  return(ww_data)
+    if (add_auxiliary_subpop) {
+      aux_subpop <- tibble::tibble(
+        "site_index" = NA,
+        "site" = NA,
+        "site_pop" = total_pop - sum(site_indices$site_pop)
+      )
+    } else {
+      aux_subpop <- tibble::tibble()
+    }
+
+    site_subpop_spine <- aux_subpop |>
+      dplyr::bind_rows(site_indices) |>
```

Comment:
iirc we discussed using the largest subpopulation (which will often be the auxiliary) as the reference by default, but also letting the user configure the reference population. As it stands, it will either be the auxiliary or the first site (whenever there is no auxiliary).

Reply:
Ok yeah, I think this configurability should be a separate issue. I think its out of scope for this PR and I'd have to think a bit about how we'd want to handle this on the user interface side.

Reply:
Sounds good. Create an issue?

Reply:
https://github.com/CDCgov/ww-inference-model/issues/196